File: src/main/java/net/danielkvasnicka/neo4jtest/NodeEntity.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package net.danielkvasnicka.neo4jtest;

import java.util.Collection;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import org.neo4j.graphdb.Direction;
import org.neo4j.graphdb.GraphDatabaseService;
import org.neo4j.graphdb.Node;
import org.neo4j.graphdb.Relationship;
import org.neo4j.graphdb.RelationshipType;
import org.neo4j.graphdb.ReturnableEvaluator;
import org.neo4j.graphdb.StopEvaluator;
import org.neo4j.graphdb.Traverser;
import org.neo4j.graphdb.Traverser.Order;

/**
 *
 * @author daniel
 */
public class NodeEntity implements Node, Map<String, Object> {

    private Node n;

    public NodeEntity(Node n) {
        this.n = n;
    }
    
    @Override
    public long getId() {
        return this.n.getId();
    }

    @Override
    public void delete() {
        this.n.delete();
    }

    @Override
    public Iterable<Relationship> getRelationships() {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public boolean hasRelationship() {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Iterable<Relationship> getRelationships(RelationshipType... rts) {
        return this.n.getRelationships(rts);
    }

    @Override
    public boolean hasRelationship(RelationshipType... rts) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Iterable<Relationship> getRelationships(Direction drctn) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public boolean hasRelationship(Direction drctn) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Iterable<Relationship> getRelationships(RelationshipType rt, Direction drctn) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public boolean hasRelationship(RelationshipType rt, Direction drctn) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Relationship getSingleRelationship(RelationshipType rt, Direction drctn) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Relationship createRelationshipTo(Node node, RelationshipType rt) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Traverser traverse(Order order, StopEvaluator se, ReturnableEvaluator re, RelationshipType rt, Direction drctn) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Traverser traverse(Order order, StopEvaluator se, ReturnableEvaluator re, RelationshipType rt, Direction drctn, RelationshipType rt1, Direction drctn1) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Traverser traverse(Order order, StopEvaluator se, ReturnableEvaluator re, Object... os) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public GraphDatabaseService getGraphDatabase() {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public boolean hasProperty(String string) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Object getProperty(String string) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Object getProperty(String string, Object o) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public void setProperty(String string, Object o) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Object removeProperty(String string) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Iterable<String> getPropertyKeys() {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Iterable<Object> getPropertyValues() {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public int size() {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public boolean isEmpty() {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public boolean containsKey(Object o) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public boolean containsValue(Object o) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Object get(Object o) {
        final String key = o.toString();
        if (this.n.hasProperty(key)) {
            return this.n.getProperty(key);
        }
        
        if (Relationships.exists(key) && 
            this.n.hasRelationship(Relationships.valueOf(key))) {
            Iterable<Relationship> relationships = this.getRelationships(Relationships.valueOf(key));
            final Iterator<Relationship> rIterator = relationships.iterator();
            Relationship first = rIterator.next();
            final NodeEntity firstNodeEntity = new NodeEntity(first.getOtherNode(this.n));
            
            if (rIterator.hasNext()) {
                Set relatedNodes = new HashSet();
                relatedNodes.add(firstNodeEntity);
                
                while(rIterator.hasNext()) {
                    relatedNodes.add(new NodeEntity(rIterator.next().getOtherNode(this.n)));
                }
                
                return relatedNodes;
            } else {
                return firstNodeEntity;
            }
        }
        
        return null;
    }
    
    @Override
    public Object put(String k, Object v) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Object remove(Object o) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public void putAll(Map<? extends String, ? extends Object> map) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public void clear() {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Set<String> keySet() {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Collection<Object> values() {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Set<Entry<String, Object>> entrySet() {
        throw new UnsupportedOperationException("Not supported yet.");
    }
    
}

File: src/main/jvm/net/danielkvasnicka/neo4jtest/NodeEntity.scala
package net.danielkvasnicka.neo4jtest

import org.neo4j.graphdb._
import java.lang.{Iterable, Object}
import scala.collection.immutable.DefaultMap
import org.neo4j.graphdb.Traverser.Order
import collection.JavaConversions._
import java.util

/**
 * Created with IntelliJ IDEA.
 * User: daniel
 * Date: 23.06.13
 * Time: 21:25
 */
class NodeEntity(n: Node) extends util.HashMap[String, Object]
    with Node {

    def getId(): Long = {
        n.getId()
    }

    def delete() {
        n.delete()
    }

    override def get(keyany: Any): Object = {
        val key = keyany.asInstanceOf[String]

        if (n.hasProperty(key)) {
            return n.getProperty(key)
        }

        if (Relationships.exists(key) &&
            n.hasRelationship(Relationships.valueOf(key))) {

            return seqAsJavaList(getRelationships(Relationships.valueOf(key)).iterator().toSeq map
                (r => new NodeEntity(r.getOtherNode(n))))
        }

        null
    }

    def getGraphDatabase: GraphDatabaseService = ???

    def hasProperty(key: String): Boolean = ???

    def getProperty(key: String): AnyRef = ???

    def getProperty(key: String, defaultValue: Any): AnyRef = ???

    def setProperty(key: String, value: Any) {}

    def removeProperty(key: String): AnyRef = ???

    def getPropertyKeys: Iterable[String] = ???

    def getPropertyValues: Iterable[AnyRef] = ???

    def getRelationships: Iterable[Relationship] = ???

    def hasRelationship: Boolean = ???

    def getRelationships(types: RelationshipType*): Iterable[Relationship] = n.getRelationships(types: _*)

    def hasRelationship(types: RelationshipType*): Boolean = ???

    def getRelationships(dir: Direction): Iterable[Relationship] = ???

    def hasRelationship(dir: Direction): Boolean = ???

    def getRelationships(`type`: RelationshipType, dir: Direction): Iterable[Relationship] = ???

    def hasRelationship(`type`: RelationshipType, dir: Direction): Boolean = ???

    def getSingleRelationship(`type`: RelationshipType, dir: Direction): Relationship = ???

    def createRelationshipTo(otherNode: Node, `type`: RelationshipType): Relationship = ???

    def traverse(traversalOrder: Order, stopEvaluator: StopEvaluator, returnableEvaluator: ReturnableEvaluator, relationshipType: RelationshipType, direction: Direction): Traverser = ???

    def traverse(traversalOrder: Order, stopEvaluator: StopEvaluator, returnableEvaluator: ReturnableEvaluator, firstRelationshipType: RelationshipType, firstDirection: Direction, secondRelationshipType: RelationshipType, secondDirection: Direction): Traverser = ???

    def traverse(traversalOrder: Order, stopEvaluator: StopEvaluator, returnableEvaluator: ReturnableEvaluator, relationshipTypesAndDirections: AnyRef*): Traverser = ???
}
